tutorials/dataframe/df106_HiggsToFourLeptons.py

```diff
@@ -0,0 +1,194 @@
+## to two Z bosons and subsequently to four leptons is called the "golden channel". The selection leads
+## to a narrow invariant mass peak on top a relatively smooth and small background, revealing
+## the Higgs at 125 GeV.
+##
+## The analysis is translated to a RDataFrame workflow processing about 300 MB of simulated events and data.
+##
+## \macro_image
+## \macro_code
+## \macro_output
+##
+## \date March 2020
+## \author Stefan Wunsch (KIT, CERN)
+
+import ROOT
+import json
+import os
+
+# Create a ROOT dataframe for each dataset
+# Note that we load the filenames from an external json file.
+path = "root://eospublic.cern.ch//eos/opendata/atlas/OutreachDatasets/2020-01-22"
+files = json.load(open(os.path.join(os.environ["ROOTSYS"], "tutorials/dataframe", "df106_HiggsToFourLeptons.json")))
+processes = files.keys()
+df = {}
+xsecs = {}
+sumws = {}
+samples = []
+for p in processes:
+    for d in files[p]:
+        # Construct the dataframes
+        folder = d[0] # Folder name
+        sample = d[1] # Sample name
+        xsecs[sample] = d[2] # Cross-section
+        sumws[sample] = d[3] # Sum of weights
+        samples.append(sample)
+        df[sample] = ROOT.RDataFrame("mini", "{}/4lep/{}/{}.4lep.root".format(path, folder, sample))
+
+# Select events for the analysis
+ROOT.gInterpreter.Declare("""
+using VecF_t = const ROOT::RVec<float>&;
+using VecI_t = const ROOT::RVec<int>&;
+bool GoodElectronOrMuon(VecI_t type, VecF_t pt, VecF_t eta, VecF_t phi, VecF_t e, VecF_t trackd0pv, VecF_t tracksigd0pv, VecF_t z0)
+{
+    for (size_t i = 0; i < type.size(); i++) {
+        ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+        if (type[i] == 11) {
+            if (pt[i] < 7000 || abs(eta[i]) > 2.47 || abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        } else {
+            if (abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        }
+    }
+    return true;
+}
+""")
+
+for s in samples:
+    # Select electron or muon trigger
+    df[s] = df[s].Filter("trigE || trigM")
+
+    # Select events with exactly four good leptons conserving charge and lepton numbers
+    df[s] = df[s].Define("good_lep", "(abs(lep_eta) < 2.5) && (lep_pt > 5000) && (lep_ptcone30 / lep_pt < 0.3) && (lep_etcone20 / lep_pt < 0.3)")\
+                 .Filter("Sum(good_lep) == 4")\
+                 .Filter("Sum(lep_charge[good_lep]) == 0")\
+                 .Define("goodlep_sumtypes", "Sum(lep_type[good_lep])")\
+                 .Filter("goodlep_sumtypes == 44 || goodlep_sumtypes == 52 || goodlep_sumtypes == 48")
+
+    # Apply additional cuts in case the lepton is an electron or muon
+    df[s] = df[s].Filter("GoodElectronOrMuon(lep_type[good_lep], lep_pt[good_lep], lep_eta[good_lep], lep_phi[good_lep], lep_E[good_lep], lep_trackd0pvunbiased[good_lep], lep_tracksigd0pvunbiased[good_lep], lep_z0[good_lep])")
```

Comment:
Maybe call this `AllElectronsAndMuonsGood`?

Reply:
agreed!